### ISSUE TITLE
Event loop task duration tracking

### ISSUE THREAD

**Issue by osi** [reactor/reactor-netty#1433]
Similar to #1398, it would be nice to be able to track the duration of event loop tasks. 

Reactor is able to do this for its schedulers via `Schedulers.enableMetrics`.

Additionally, it would be nice to set a threshold for the duration of any given task, and be able to log an event with details about that task (for identifying specific instances in the metrics)

**Comment by kevinten10** [comment]
Full monitoring may cause redundant information and excessive performance loss.

Is it more reasonable to perform sampling of this part of the function?

like: `sample(() -> Schedulers.enableMetrics,
       (Predicate) () -> { /* Trigger threshold */ });`

**Comment by osi** [comment]
Sampling may be fine. My goal is to identify work taking place on an event loop thread that is effectively blocking the loop due to how long it takes. (even if the work is fully non-blocking).

**Comment by mateusz-nalepa** [comment]
Hey @violetagg !
I was thinking about adding something like this to the class `SingleThreadEventExecutor`

```java
private void execute(Runnable task, boolean immediate) {
     MeteredRunnable meteredTask = new MeteredRunnable(task, threadProperties.name());

...
class MeteredRunnable implements Runnable {

    private final Runnable task;
    private final String executorName;
    private final Long start;

    MeteredRunnable(Runnable task, String executorName) {
        this.task = task;
        this.executorName = executorName;
        this.start = System.nanoTime();
    }

    @Override
    public void run() {
        Duration duration = Duration.ofNanos(System.nanoTime() - start);
        System.out.println("Runnable took: " + duration + " for executor name: " + executorName);
        task.run();
    }
}
```

However, it turns out that class `SingleThreadEventExecutor` is located in https://github.com/netty/netty that has no access to Micrometer.
So I thought about some lambda in class `DefaultLoopResources` which could be passed down to `SingleThreadEventExecutor`, but I'm not sure about this approach :/ 

I'm open to your recommendations on how to add this metric! Right now, I'm out of ideas :D

**Comment by violetagg** [comment]
@mateusz-nalepa I haven't spent time for researching this and how to be implemented so that we have both metrics and the performance is still intact.
There are some experiments, check linked issues here, also https://github.com/netty/netty/pull/11293, https://github.com/netty/netty/pull/9080

**Comment by mateusz-nalepa** [comment]
Hi @violetagg 

I've failed with preparing this metric, but I was able to prepare workaround for `part of it`, I mean:

`How long I'm waiting for a WebClient thread, when ReactorClientHttpConnector(default) is used in a Spring Boot app?`

So basically I'm very happy, but still I would say, that it would be good to have event loop task duration tracking metric out of the box. 

Overall, I feel, that this is not related only to Netty, but probably to every place, where tasks can wait in thread pool. Why? When response times are increasing then sometimes** it's hard to tell, if this is something related with some external service, or just tasks waiting patiently in a some queue. For example: a lot of serialization/deserialization is being done by default on WebClient threads, and it should be done on some parallel threads. But it's good to know when to switch to another thread pool. Why? When `Pending Request Time` is near zero, then switching to another thread pool adds only excessive context switch.

**HttpClient current metrics will return correct values, but where are missing seconds, when response times for endpoint are higher? Thanks to task duration tracking we can have it.

So please let me know what do you think about this, I'm open to your feedback as always.

### Workaround

`Spring WebClient + ExchangeFilterFunction` and `HttpClient.doOnRequest` works fine

Reproducible example is here https://github.com/mateusz-nalepa/pending-task-time
All information related how to run are in readme file. I've included also simulation scenario in the form of an image.

TL;DR
Thanks to log line `End WEB_CLIENT_PENDING_REQUEST_TIME. Took: PT X s` I know, that WebClient tasks are waiting in a queue for a thread, so I can use `.publishOn()` operator to move processing to another thread after getting response from WebClient. In simulation scenario it lowers endpoint response times from 25s to 17s, cause 8s is just waiting.

**Comment by mateusz-nalepa** [comment]
Hi @violetagg 
It turns out, that Spring probably has a bug. I'm waiting for info from Spring Team
https://github.com/spring-projects/spring-framework/issues/35302

So from my side, probably fixing this bug is good enough
